Allow attack anims/overlays to start before weapons fire

reaperrr opened the pull request:
This allows attack animations/overlays to ignore `Armament.FireDelay` and start right away, which in turn allows fixing Grenadier throw anim timing.

Closes #11331.

reaperrr committed:
Move WithAttackOverlay to Mods.Common

reaperrr committed:
Introduce INotifyAttack.PreparingAttack

And use it to allow triggering attack animations/overlays at a different point
than the actual weapon firing.

reaperrr committed:
Add AttackSounds trait

This trait plays sounds when preparing to attack or attacking.

reaperrr committed:
Make all Attack* traits listen to Stop order

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -35,6 +35,7 @@ public interface IPlaceBuildingDecorationInfo : ITraitInfo
 	}
 
 	public interface INotifyAttack { void Attacking(Actor self, Target target, Armament a, Barrel barrel); }
+	public interface INotifyAttackAnimation { void AttackingAnimation(Actor self, Target target, Armament a, Barrel barrel); }
```
Can we get a `[RequireExplicitImplementation]` here?

reaperrr replied:
Done. That caught the implicit implementation in `WithInfantryBody` as well.

reaperrr commented:
Updated.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Sound/WithAttackAnimationSounds.cs:
```diff
@@ -0,0 +1,42 @@
+	public class WithAttackAnimationSoundsInfo : ITraitInfo
+	{
+		[Desc("Play a randomly selected sound from this list along with any attack animations or overlays.")]
+		public readonly string[] Sounds = null;
```
Make it empty, not null.

reaperrr commented:
Updated.

reaperrr commented:
Updated.

Replaced the `AttackAnimsIgnoreFireDelay` boolean with a more straightforward and flexible `AttackAnimationDelay` that defaults to `FireDelay`.
Although it's unlikely to see much use, you may never know when some 3rd-party mod wants to delay both, but not to the same tick, and this way we can avoid having to write upgrade rules and upgrade our mods later on.

pchote commented:
Rewording my thoughts from IRC to be a bit clearer:

I'm thinking we could make the abstraction a bit cleaner by merging `INotifyAttackAnimation` into `INotifyAttack` by defining a `INotifyAttack.PreparingAttack` method that is called immediately when preparaing the armament.  The `WithAttackAnimation` trait can then have its own properties for defining the delay and an enum to say whether that delay is relative to preparing or the actual attack.  Likewise for the attack sounds

This brings two advantages:
1) Armaments don't have to concern themselves with animation-specific code.
2) Different downstream traits (e.g. animations vs sounds) can use different delays.

reaperrr commented:
Updated, done as suggested.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -47,13 +54,25 @@ public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+		{
+			if (info.DelayRelativeTo == AttackDelayType.Attack && !string.IsNullOrEmpty(info.AttackSequence))
+				self.World.AddFrameEndTask(w => w.Add(new DelayedAction(info.Delay, () =>
+					{
```
One level of indentation too much.

reaperrr replied:
Hm, are you sure? the brackets belong to the `DelayedAction`, not the `AddFrameEndTask`.

pchote replied:
Cases like this are why we have a(n often ignored) rule to avoid hanging blocks.  When you have an inner block with braces all outer blocks should also have braces, so please add them here.

reaperrr commented:
Updated, added brackets for better readability.

obrakmann commented on OpenRA.Mods.Common/Traits/Sound/PreparingAttackSounds.cs:
```diff
@@ -0,0 +1,45 @@
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits.Render
```
That doesn't look like the right place.

reaperrr replied:
Ouch, well spotted.

obrakmann commented:
This removes the burst-y firing behaviour from the obelisk. Are people cool with that?

obrakmann commented:
Nevermind, seems that I misremembered the obelisk firing in two-shot bursts. :+1: then, after the namespace fixup.

reaperrr commented:
Fixed.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -57,19 +57,29 @@ public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+				self.World.AddFrameEndTask(w =>
+				{
+					w.Add(new DelayedAction(info.Delay, () =>
 					{
```
These brackets actually aren't necessary, this is a single instruction.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -57,19 +57,29 @@ public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+				self.World.AddFrameEndTask(w =>
+				{
+					w.Add(new DelayedAction(info.Delay, () =>
 					{
```
Same here.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Sound/PreparingAttackSounds.cs:
```diff
@@ -0,0 +1,45 @@
+
+		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel)
+		{
+			if (info.Sounds != null && info.Sounds.Any())
```
`info.Sounds` won't be null.

reaperrr commented:
Updated.

pchote commented on mods/cnc/rules/structures.yaml:
```diff
@@ -757,15 +757,15 @@ OBLI:
-		ChargeAudio: obelpowr.aud
-		ReloadDelay: 40
-		InitialChargeDelay: 50
+		FireDelay: 40
```
@GraionDilach pointed out in IRC that this makes the actor ignore the maximum range check, and fire unconditionally on the target.  `AttackCharge` (actually `ChargeFire`) has a specific extra range check that avoids that.

pchote replied:
Fast units can move quite a way (4-5 cells) past the max range during the charge animation, so this is a pretty bad regression.

reaperrr replied:
> so this is a pretty bad regression.

I agree, but I'd like to point out that the underlying problem is (or was, as I have a fix for it) that `Armament` schedules the delayed action almost unconditionally, a larger `FireDelay` for other weapons would have the same issue, ignoring the minimum range as well.

pchote replied:
Indeed, but in most (all?) other existing cases the `FireDelay`s are small enough for this to be much less of an issue.

reaperrr commented:
Updated.

Projectiles are now only launched if the target didn't move out of max range or inside min range during the `FireDelay`.

pchote commented:
I'm not convinced that's a good approach.  We've already committed to firing the armament by this point, so I think it would be better to change the target position to the ground at the edge of the max range.  Obelisk is a special case that is still best handled by its own targeting trait (ref the recent discussions in IRC).

Nelax commented:
I know at least what I plan to use this logic for, a turreted charging laser weapon, that would not be preferable but I could see it being the preferred way of firing for any normal projectile weapon.

pchote commented:
Can we, for now, drop f6bca6c and the obelisk and tesla changes?

This is otherwise a straightforward and uncontroversial plumbing improvement that doesn't need to be blocked by these other issues.

reaperrr commented:
> Can we, for now, drop f6bca6c and the obelisk and tesla changes?

Okay, done.

Phrohdoh commented:
59816017730d49697dac5d4c7123db07d4be8fd9 should be squashed into whichever commit introduces that code.

reaperrr committed:
Add start delay functionality to attack animation traits

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -47,13 +54,35 @@ public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+		{
+			if (info.DelayRelativeTo == AttackDelayType.Attack && !string.IsNullOrEmpty(info.AttackSequence))
+			{
+				self.World.AddFrameEndTask(w =>
```
Can you please pull out a `void PlayAnimation` method to avoid this duplication?

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAttackOverlay.cs:
```diff
@@ -52,10 +57,34 @@ public WithAttackOverlay(ActorInitializer init, WithAttackOverlayInfo info)
+		{
+			if (info.DelayRelativeTo == AttackDelayType.Attack)
+			{
+				self.World.AddFrameEndTask(w =>
```
Same here.

pchote commented on OpenRA.Mods.Common/Traits/Sound/PreparingAttackSounds.cs:
```diff
@@ -0,0 +1,71 @@
+		{
+			if (info.DelayRelativeTo == AttackDelayType.Attack && info.Sounds.Any())
+			{
+				self.World.AddFrameEndTask(w =>
```
Same here.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -34,7 +36,12 @@ public interface IPlaceBuildingDecorationInfo : ITraitInfo
 	}
 
-	public interface INotifyAttack { void Attacking(Actor self, Target target, Armament a, Barrel barrel); }
+	public interface INotifyAttack
```
Can you please add `[RequireExplicitImplementation]` to this?

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackOmni.cs:
```diff
@@ -29,6 +29,14 @@ public override Activity GetAttackActivity(Actor self, Target newTarget, bool al
 			return new SetTarget(this, newTarget);
 		}
 
+		public override void ResolveOrder(Actor self, Order order)
```
What's the motivation for moving this from `AttackCharge` to `AttackOmni` (why not `AttackBase`)?

reaperrr replied:
I just wasn't sure if there's a reason why this was not part of `AttackBase` and didn't want to take any risks, but I guess it can't really hurt to move it to `AttackBase`.

pchote commented on OpenRA.Mods.Common/Traits/Infantry/ScaredyCat.cs:
```diff
@@ -82,12 +82,14 @@ public void Damaged(Actor self, AttackInfo e)
 		}
 
-		public void Attacking(Actor self, Target target, Armament a, Barrel barrel)
+		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
```
Does anyone know what the purpose of this is?  It was introduced in 11c45e43884f57de6daa2a38543eb7670a0c99e4 but doesn't appear to be used at all.

reaperrr replied:
The idea is (or was) that a scared actor would have a certain probability of panicking when attacking, I guess (in the original games, civilians that were shot at would run around in panic between attacks even while attacking a target). Not sure if that still works, but I consider that out of scope for this PR either way.

reaperrr commented:
Updated.

Everything should work now, but I guess this changed too much since @obrakmann reviewed it so I'm removing the Needs +2.

obrakmann commented:
still looks fine to me, :+1:

reaperrr commented:
Rebased so reviewers won't have to switch dependencies between this and newer PRs.

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/579de74a8ace61b9dbe2fcabcfadc62349f5df7f...36162753843e86ccd04d9faebe93e46b05c9ef2d)
